chainerio/__init__.py

```diff
@@ -68,60 +87,194 @@ def open(file_path: str, mode: str = 'rb',
 
 
-def set_root(uri_or_handler: str) -> None:
+def set_root(uri_or_handler: Union[str, Type['IO']]) -> None:
+    """Sets the current context to ``uri_or_handler``
+
+    The context here refers to the default handler and the root path.
+    The default handler points to a filesystem or a container which
+    ChainerIO uses when called without a full URI. Overrides by the full URI.
+    The default handler can be created by :func:`create_handler` with
+    the name of the handler.
+
+    Example::
+
+        # Case 1
+        # set_root by the name of uri
+        chainerio.set_root("posix")
+        # open a file on posix filesystem with path "some/file"
+        chainerio.open("some/file")
+        # override with a full uri
+        chainerio.open("hdfs:///some/file/on/hdfs")
+
+        # Case 2
+        # set_root by uri
+        chainerio.set_root("hdfs:///some/directory")
+        # open a file "on/hdfs" on hdfs under the " some/directory"
+        chainerio.open("on/hdfs")
+
+        # Case 3
+        handler = chainerio.create_handler('hdfs')
+        # set_root by handler
+        chainerio.set_root(handler)
+        # open a file "/some/file/on/hdfs" on hdfs
+        chainerio.open("/some/file/on/hdfs")
+
+        # Case 4
+        handler = chainerio.open_as_container('some.zip')
+        # set_root by handler
+        chainerio.set_root(handler)
+        # open a file "img.jpg" in "some.zip"
+        chainerio.open("img.jpg")
+
+    The root path refers to a directory that ChainerIO works on.
+    It is similar to current working directory, ``CWD`` in terms of the shell
+    environment.
+    The root path will only be set
+    when the ``uri_or_handler`` points to a directory.
+    Otherwise, it will be set to default,
+    which represents the default working directory as follow:
+
+    +---------+---------------------------+
+    |         |     Default Root Path     |
+    +=========+===========================+
+    | POSIX   | current working directory |
+    +---------+---------------------------+
+    | HDFS    |     /user/``USERNAME`` #1 |
+    +---------+---------------------------+
+    | zip     |      top directory        |
+    +---------+---------------------------+
+
+    #1 For the details about the ``USERNAME`` in HDFS, please refer to
+    `HDFS Document <https://hadoop.apache.org/docs/current/\
+            hadoop-project-dist/hadoop-hdfs/HdfsPermissionsGuide.html>`_
+
+
+    Args:
+        uri_or_handler (str or :class:~`chainerio.IO`): The ``uri_or_handler``
+            can accept the following three kinds of values:
+
+            1. the name of a handler (string): set the default handler
+                to the corresponding handler, and root path.
+                See :func:`create_handler` for supported name of handlers.
+
+            2. an uri of directory (string): set the context
+                to use the corresponding handler and set
+                the root path to the given directory.
```

Comment:
This case seems not working
```
>>> chainerio.set_root("hdfs://tmp")
hdfsGetPathInfo(hdfs://tmp): getFileInfo error:
IllegalArgumentException: Wrong FS: hdfs://tmp, expected: hdfs://mn1s-nameservice1java.lang.IllegalArgumentException: Wrong FS: hdfs://tmp, expected: hdfs://mn1s-nameser
vice1
        at org.apache.hadoop.fs.FileSystem.checkPath(FileSystem.java:662)
        at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:222)
        at org.apache.hadoop.hdfs.DistributedFileSystem.access$000(DistributedFileSystem.java:114)
        at org.apache.hadoop.hdfs.DistributedFileSystem$20.doCall(DistributedFileSystem.java:1266)
        at org.apache.hadoop.hdfs.DistributedFileSystem$20.doCall(DistributedFileSystem.java:1262)
        at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
        at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1262)
        at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1418)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/vol10/kota/src/chainerio/chainerio/__init__.py", line 75, in set_root
    default_context.set_root(uri_or_handler)
  File "/mnt/vol10/kota/src/chainerio/chainerio/_context.py", line 84, in set_root
    if not handler.isdir(self._root):
  File "/mnt/vol10/kota/src/chainerio/chainerio/filesystems/hdfs.py", line 153, in isdir
    stat = self.stat(file_path)
  File "/mnt/vol10/kota/src/chainerio/chainerio/filesystems/hdfs.py", line 139, in stat
    return self.connection.stat(path)
  File "pyarrow/io-hdfs.pxi", line 247, in pyarrow.lib.HadoopFileSystem.stat
  File "pyarrow/error.pxi", line 83, in pyarrow.lib.check_status
pyarrow.lib.ArrowIOError: Calling GetPathInfo for hdfs://tmp failed. errno: 22 (Invalid argument)
```

Reply:
Seems ChainerIO freaked out after the error above
```
>>> chainerio.get_root_dir()
'hdfs://tmp'
>>> chainerio.open("hello.txt", "wb")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/vol10/kota/src/chainerio/chainerio/__init__.py", line 68, in open
    errors, newline, closefd, opener)
  File "/mnt/vol10/kota/src/chainerio/chainerio/io.py", line 19, in wrapper
    errors, newline, closefd, opener)
  File "/mnt/vol10/kota/src/chainerio/chainerio/filesystems/posix.py", line 25, in open
    newline, closefd, opener)
FileNotFoundError: [Errno 2] No such file or directory: 'hdfs://tmp/hello.txt'
```

Reply:
I think you made a mistake here: "hdfs://tmp" should be "hdfs:///tmp" as stated in the error message: `IllegalArgumentException: Wrong FS: hdfs://tmp, expected: hdfs://mn1s-nameservice1java.lang.IllegalArgumentException: Wrong FS: hdfs://tmp, expected: hdfs://mn1s-nameser
vice1`

```
>>> chainerio.set_root("hdfs:///tmp/")
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/usr/lib/zookeeper/lib/slf4j-log4j12-1.7.5.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/usr/share/java/slf4j-simple.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
19/09/26 17:34:29 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
>>> list(chainerio.list())
['.cloudera_health_monitoring_canary_files', '42102068-6b80-4089-9b37-f0338c4f9a2f.parquet', 'a', 'b924cb32-856c-4b92-bb5b-b5a4796d549e.parquet', 'hadoop-yarn', 'hbase-staging', 'xxx-spark', 'hello', 'hive', 'logs']
>>>
```
But indeed ChainerIO should handle this error